src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsStreamIterator.java

```diff
@@ -0,0 +1,240 @@
+import com.google.common.base.Predicate;
+import com.google.common.collect.ForwardingIterator;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+
+/**
+ * An iterator for streaming genomic data via gRPC with shard boundary semantics.
+ * 
+ * Includes complex retry logic to upon failure resume the stream at the last known good start
+ * position without returning duplicate data.
+ * 
+ * @param <A> Streaming request type.
+ * @param <B> Streaming response type.
+ * @param <C> Genomic data type returned by stream.
+ * @param <D> Blocking stub type.
+ */
+public abstract class GenomicsStreamIterator<A, B, C, D extends io.grpc.stub.AbstractStub<D>> extends ForwardingIterator<B> {
+  private static final Logger LOG = Logger.getLogger(GenomicsStreamIterator.class.getName());
+  private final ExponentialBackOff backoff;
+  protected final D stub;
+  protected final GenomicsChannel genomicsChannel;
+  protected final A originalRequest;
+
+  // For client-side enforcement of a strict shard boundary.
+  protected Predicate<C> shardPredicate;
+
+  // Stateful members used to facilitate complex retry behavior for gRPC streams.
+  private Iterator<B> delegate;
+  private C lastSuccessfulDataItem;
+  private String idSentinel;
+
+  /**
+   * Create a stream iterator that can enforce shard boundary semantics and perform retries.
+   * 
+   * @param request The request for the shard of data.
+   * @param auth The OfflineAuth to use for the request.
+   * @param shardBoundary The shard boundary semantics to enforce.
+   * @param fields Which fields to include in a partial response or null for all. NOT YET
+   *        IMPLEMENTED.
+   * @throws IOException
+   * @throws GeneralSecurityException
+   */
+  public GenomicsStreamIterator(A request, GenomicsFactory.OfflineAuth auth,
+      ShardBoundary.Requirement shardBoundary, String fields) throws IOException,
+      GeneralSecurityException {
+    this.originalRequest = request;
+    genomicsChannel = GenomicsChannel.fromOfflineAuth(auth);
+    stub = createStub(genomicsChannel);
+
+    // Using default backoff settings. For details, see
+    // https://developers.google.com/api-client-library/java/google-http-java-client/reference/1.19.0/com/google/api/client/util/ExponentialBackOff
+    backoff = new ExponentialBackOff.Builder().build();
+
+    // RETRY STATE: Initialize settings.
+    delegate = createIterator(originalRequest);
+    lastSuccessfulDataItem = null;
+    idSentinel = null;
+  }
+
+  abstract D createStub(GenomicsChannel genomicsChannel);
+
+  abstract Iterator<B> createIteratorFromStub(A request);
+
+  abstract long getRequestStart(A streamRequest);
+
+  abstract long getDataItemStart(C dataItem);
+
+  abstract String getDataItemId(C dataItem);
+
+  abstract A getRevisedRequest(long updatedStart);
+
+  abstract List<C> getDataList(B response);
+
+  abstract B setDataList(B response, Iterable<C> dataList);
```

Comment:
This naming threw me off originally. This is not actually a standard Java setter. I would suggest calling this `buildResponse()` instead.